keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -532,6 +534,45 @@ class SqlArtifactRepository(
     return vetoes.values.toList()
   }
 
+  /**
+   *  Record that an artifact version should be vetoed (marked as bad) in a target environment
+   *
+   *  @param deliveryConfig config object for looking up artifact information by artifact reference
+   *  @param veto information about the artifact version to be vetoed and the target environment to veto in
+   *  @param force if true, record the version as vetoed even if it is marked as the promotion reference
+   *               (automated rollback target) of a previously vetoed version
+   *
+   *  @return true on success
+   *
+   * Preconditions for success:
+   *
+   *  1. [veto.version] is not currently pinned in the target environment
+   *  2. There exists a record R in the environment_artifacts_versions table such that:
+   *     a. R references the artifact version to veto and the target environment encoded in [veto]
```

Comment:
Note that precondition 2 (specifically, 2a) is what's leading to #1086, which happens when there is no record in the `environment_artifacts_versions` table that corresponds to the vetoed version.